import sys, os, json, socket, threading
from PySide6.QtWidgets import (
    QApplication, QStackedWidget, QPushButton, QLineEdit, QMessageBox,
    QListWidget, QListWidgetItem, QTextEdit, QLabel, QWidget, QHBoxLayout, QToolButton,
    QFileDialog
)
from PySide6.QtUiTools import QUiLoader
from PySide6.QtCore import QFile, QIODevice, QObject, Signal, Qt
from PySide6.QtGui import QIcon
from PySide6.QtWidgets import QAbstractItemView
import ipaddress
import re
from PySide6.QtGui import QDesktopServices
from PySide6.QtCore import QUrl
from datetime import datetime

ui_path = r"D:\0_Computer_Network_Course_Design\second_client\client.ui"
icon_path = r"D:\0_Computer_Network_Course_Design\second_client\icon.ico"
eye_open = r"D:\0_Computer_Network_Course_Design\second_client\eye_open.ico"
eye_close = r"D:\0_Computer_Network_Course_Design\second_client\eye_close.ico"

def send_request(server_ip, server_port, req):
    with socket.create_connection((server_ip, server_port), timeout=5) as sock:
        sock.sendall((json.dumps(req, ensure_ascii=False) + "\n").encode("utf-8"))
        buf = bytearray()
        while True:
            chunk = sock.recv(4096)
            if not chunk:
                raise ConnectionError("æœåŠ¡ç«¯æ–­å¼€è¿æ¥")
            buf.extend(chunk)
            i = buf.find(b"\n")
            if i != -1:
                line = bytes(buf[:i]).decode("utf-8", errors="replace")
                return json.loads(line)

def getLocalIp():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("8.8.8.8", 80))
        return s.getsockname()[0]
    except Exception:
        return "127.0.0.1"
    finally:
        s.close()

def recv_line(conn: socket.socket, buf: bytearray):
    while True:
        i = buf.find(b"\n")
        if i != -1:
            line = bytes(buf[:i])
            del buf[:i + 1]
            return line.decode("utf-8", errors="replace")
        chunk = conn.recv(4096)
        if not chunk:
            return None
        buf.extend(chunk)

def recv_exact(conn: socket.socket, size: int):
    data = b""
    while len(data) < size:
        chunk = conn.recv(min(4096, size - len(data)))
        if not chunk:
            raise ConnectionError("æ–‡ä»¶æ¥æ”¶ä¸­æ–­")
        data += chunk
    return data


def send_json(conn: socket.socket, obj: dict):
    conn.sendall((json.dumps(obj, ensure_ascii=False) + "\n").encode("utf-8"))

class UiEmitter(QObject):
    users_update = Signal(list)            # list[dict]
    msg_in = Signal(str, str, object)         # from, to, text
    info = Signal(str)

class MainWindow:
    def __init__(self):
        self.sock = None
        self.recv_thread = None
        self.stop_recv = threading.Event()

        self.username = ""
        self.local_ip = getLocalIp()

        # peer -> {"messages":[(is_me,text)], "unread":int, "online":bool, "ip":str}
        self.sessions = {}
        self.current_peer = None

        self.emitter = UiEmitter()
        self.load_ui()
        self.bind_signals()

        self.emitter.users_update.connect(self.on_users_update)
        self.emitter.msg_in.connect(self.on_message_in)
        self.emitter.info.connect(lambda s: self.popUp("æç¤º", s, ok=True))

        self.last_msg_time = None

    def load_ui(self):
        loader = QUiLoader()
        file = QFile(ui_path)
        file.open(QIODevice.ReadOnly)
        self.window = loader.load(file)
        file.close()

        self.window.setWindowTitle("å®¢æˆ·ç«¯")
        self.window.setWindowIcon(QIcon(icon_path))

        self.interfaceWidget = self.window.findChild(QStackedWidget, "interfaceWidget")
        self.loginButton = self.window.findChild(QPushButton, "loginButton")
        self.enrollButton = self.window.findChild(QPushButton, "enrollButton")

        self.userNameEdit = self.window.findChild(QLineEdit, "userNameEdit")
        self.userPwdEdit = self.window.findChild(QLineEdit, "userPwdEdit")

        self.enrollUserNameEdit = self.window.findChild(QLineEdit, "enrollUserNameEdit")
        self.enrollUserPwdEdit = self.window.findChild(QLineEdit, "enrollUserPwdEdit")
        self.enrollSummitButton = self.window.findChild(QPushButton, "enrollSummitButton")
        self.enrollServerIPEdit = self.window.findChild(QLineEdit, "enrollServerIPEdit")
        self.enrollServerPortEdit = self.window.findChild(QLineEdit, "enrollServerPortEdit")

        self.serverIPEdit = self.window.findChild(QLineEdit, "serverIPEdit")
        self.serverPortEdit = self.window.findChild(QLineEdit, "serverPortEdit")

        self.loginEyeButton = self.window.findChild(QToolButton, "loginEyeButton")
        if self.userPwdEdit:
            self.userPwdEdit.setEchoMode(QLineEdit.Password)
        if self.loginEyeButton:
            self.loginEyeButton.setCheckable(True)
            self.loginEyeButton.setIcon(QIcon(eye_close))

        # èŠå¤©é¡µæ§ä»¶ï¼ˆindex=2ï¼‰
        self.chatListWidget = self.window.findChild(QListWidget, "chatListWidget")
        self.messageListWidget = self.window.findChild(QListWidget, "messageListWidget")
        self.inputTextEdit = self.window.findChild(QTextEdit, "inputTextEdit")
        self.sendMsgButton = self.window.findChild(QPushButton, "sendMsgButton")
        self.sendFileButton = self.window.findChild(QPushButton, "sendFileButton")

        self.messageListWidget.setSpacing(3)

        # å·¦ä¾§ç”¨æˆ·åˆ—è¡¨ï¼šé€‰ä¸­é«˜äº®ï¼ˆQSSï¼‰
        if self.chatListWidget:
            self.chatListWidget.setSelectionMode(QAbstractItemView.SingleSelection)
            self.chatListWidget.setStyleSheet("""
                QListWidget {
                    background: #FFFFFF;
                    border: 1px solid #E5E5E5;
                    outline: 0;
                }
                QListWidget::item {
                    padding: 10px 8px;
                    margin: 2px 6px;
                    border-radius: 8px;
                }
                QListWidget::item:hover {
                    background: #F2F6FF;
                }
                QListWidget::item:selected {
                    background: #DCEBFF;
                    color: #003A8C;
                    font-weight: 600;
                }
            """)

        self.messageListWidget.itemDoubleClicked.connect(self.open_file)

    def bind_signals(self):
        self.loginButton.clicked.connect(self.do_login)
        self.enrollButton.clicked.connect(lambda: self.interfaceWidget.setCurrentIndex(1))
        self.enrollSummitButton.clicked.connect(self.do_register)

        if self.sendMsgButton:
            self.sendMsgButton.clicked.connect(self.send_chat_message)
        if self.chatListWidget:
            self.chatListWidget.itemClicked.connect(self.on_peer_clicked)
        if self.sendFileButton:
            self.sendFileButton.clicked.connect(self.send_file)

        if self.loginEyeButton and self.userPwdEdit:
            self.loginEyeButton.toggled.connect(
                lambda checked: self.toggle_password(self.userPwdEdit, self.loginEyeButton, checked))

    def popUp(self, title, msg, ok=True):
        if ok:
            QMessageBox.information(self.window, title, msg)
        else:
            QMessageBox.warning(self.window, title, msg)

    def isValidIp(self, IP):
        s = (IP or "").strip()
        try:
            return ipaddress.ip_address(s).version == 4
        except ValueError:
            return False

    def is_valid_username(self, username: str):
        username = (username or "").strip()
        if not username:
            return False, "ç”¨æˆ·åä¸èƒ½ä¸ºç©º"
        # ä¸èƒ½ä»¥æ•°å­—å¼€å¤´
        if username[0].isdigit():
            return False, "ç”¨æˆ·åä¸èƒ½ä»¥æ•°å­—å¼€å¤´"
        return True, ""

    def is_strong_password(self, password: str):
        password = password or ""
        if len(password) < 6:
            return False, "å¯†ç é•¿åº¦ä¸èƒ½å°‘äº6ä½"
        if not re.search(r"[A-Za-z]", password):
            return False, "å¯†ç å¿…é¡»åŒ…å«å­—æ¯"
        if not re.search(r"\d", password):
            return False, "å¯†ç å¿…é¡»åŒ…å«æ•°å­—"
        # ç‰¹æ®Šå­—ç¬¦ï¼šåªè¦ä¸æ˜¯å­—æ¯æ•°å­—å³å¯
        if not re.search(r"[^A-Za-z0-9]", password):
            return False, "å¯†ç å¿…é¡»åŒ…å«ç‰¹æ®Šå­—ç¬¦"
        return True, ""

    def do_register(self):
        userName = (self.enrollUserNameEdit.text() if self.enrollUserNameEdit else "")
        password = (self.enrollUserPwdEdit.text() if self.enrollUserPwdEdit else "")
        serverIP = (self.enrollServerIPEdit.text() if self.enrollServerIPEdit else "")
        serverPort = int((self.enrollServerPortEdit.text() if self.enrollServerPortEdit else "0") or "0")

        if not self.isValidIp(serverIP):
            self.popUp("æ³¨å†Œå¤±è´¥","æœåŠ¡å™¨IPæ ¼å¼ä¸åˆæ³•", ok=False)

        ok_u, msg_u = self.is_valid_username(userName)
        if not ok_u:
            self.popUp("æ³¨å†Œå¤±è´¥", msg_u, ok=False)
            return

        ok_p, msg_p = self.is_strong_password(password)
        if not ok_p:
            self.popUp("æ³¨å†Œå¤±è´¥", msg_p, ok=False)
            return

        try:
            resp = send_request(serverIP, serverPort, {
                "action": "register",
                "username": userName,
                "password": password
            })
            self.popUp("æ³¨å†Œç»“æœ", resp.get("msg", ""), ok=resp.get("ok", False))
            if resp.get("ok"):
                self.interfaceWidget.setCurrentIndex(0)
        except Exception as e:
            self.popUp("ç½‘ç»œé”™è¯¯ï¼Œè¯·æ£€æŸ¥æœåŠ¡å™¨IP", str(e), ok=False)

    def do_login(self):
        username = (self.userNameEdit.text() if self.userNameEdit else "").strip()
        password = (self.userPwdEdit.text() if self.userPwdEdit else "")
        serverIP = (self.serverIPEdit.text() if self.serverIPEdit else "").strip()
        serverPort = int((self.serverPortEdit.text() if self.serverPortEdit else "0") or "0")

        if not self.isValidIp(serverIP):
            self.popUp("ç™»å½•å¤±è´¥","æœåŠ¡å™¨IPæ ¼å¼ä¸åˆæ³•", ok=False)
            return

        if not username or not password:
            self.popUp("ç™»å½•å¤±è´¥", "ç”¨æˆ·å/å¯†ç ä¸èƒ½ä¸ºç©º", ok=False)
            return

        try:
            resp = send_request(serverIP, serverPort, {
                "action": "login",
                "username": username,
                "password": password,
                "client_ip": self.local_ip
            })
            self.popUp("ç™»å½•ç»“æœ", resp.get("msg", ""), ok=resp.get("ok", False))

            if resp.get("ok"):
                self.username = username
                self.interfaceWidget.setCurrentIndex(2)
                self.start_chat_connection(serverIP, serverPort)

        except Exception as e:
            self.popUp("ç½‘ç»œé”™è¯¯ï¼Œè¯·æ£€æŸ¥æœåŠ¡å™¨IP", str(e), ok=False)

    def toggle_password(self, edit: QLineEdit, btn: QToolButton, checked: bool):
        """
        checked=True  -> æ˜¾ç¤ºæ˜æ–‡
        checked=False -> éšè—å¯†ç 
        """
        if checked:
            edit.setEchoMode(QLineEdit.Normal)
            btn.setIcon(QIcon(eye_open))
        else:
            edit.setEchoMode(QLineEdit.Password)
            btn.setIcon(QIcon(eye_close))

        # ä¿æŒå…‰æ ‡åœ¨æœ«å°¾ï¼ˆä½“éªŒæ›´åƒå¾®ä¿¡/QQï¼‰
        edit.setFocus()
        edit.setCursorPosition(len(edit.text()))

    # ====== é•¿è¿æ¥ ======
    def start_chat_connection(self, server_ip, server_port):
        self.close_chat_connection()
        try:
            self.sock = socket.create_connection((server_ip, server_port), timeout=5)
            self.sock.settimeout(None)
            self.stop_recv.clear()

            # ä¸Šçº¿
            send_json(self.sock, {
                "action": "online",
                "username": self.username,
                "client_ip": self.local_ip
            })

            self.recv_thread = threading.Thread(target=self.recv_loop, daemon=True)
            self.recv_thread.start()

            self.emitter.info.emit(f"å·²è¿æ¥ {server_ip}:{server_port}  æœ¬æœºIP={self.local_ip}")
        except Exception as e:
            self.popUp("è¿æ¥å¤±è´¥", str(e), ok=False)

    def close_chat_connection(self):
        try:
            self.stop_recv.set()
            if self.sock:
                try:
                    self.sock.shutdown(socket.SHUT_RDWR)
                except Exception:
                    pass
                self.sock.close()
        finally:
            self.sock = None

    def recv_loop(self):
        buf = bytearray()
        try:
            while not self.stop_recv.is_set():
                line = recv_line(self.sock, buf)
                if line is None:
                    break

                try:
                    msg = json.loads(line)
                except Exception:
                    continue

                action = msg.get("action")

                if action == "user_state_list":
                    users = msg.get("users", [])
                    self.emitter.users_update.emit(users)

                elif action == "msg":
                    _from = msg.get("from", "")
                    _to = msg.get("to", "")
                    text = msg.get("text", "")
                    self.emitter.msg_in.emit(_from, _to, text)


                elif action == "file_meta":
                    _from = msg.get("from", "")
                    filename = msg.get("filename", "")
                    filesize = int(msg.get("filesize", 0))
                    save_path = os.path.join(
                        os.path.expanduser("~/Downloads"), filename
                    )
                    try:
                        remaining = filesize
                        file_data = b""

                        # â­ å…ˆæ¶ˆè€— buf é‡Œçš„æ®‹ç•™ï¼ˆéå¸¸é‡è¦ï¼‰
                        if len(buf) > 0:
                            take = min(len(buf), remaining)
                            file_data += bytes(buf[:take])
                            del buf[:take]
                            remaining -= take

                        # â­ å†ä» socket æ”¶å‰©ä½™
                        while remaining > 0:
                            chunk = self.sock.recv(min(4096, remaining))
                            if not chunk:
                                raise ConnectionError("æ–‡ä»¶æ¥æ”¶ä¸­æ–­")
                            file_data += chunk
                            remaining -= len(chunk)

                        with open(save_path, "wb") as f:
                            f.write(file_data)

                        self.emitter.msg_in.emit(
                            _from,
                            self.username,
                            {
                                "type": "file",
                                "filename": filename,
                                "path": save_path,
                                "is_me": False
                            }
                        )
                    except Exception as e:
                        self.emitter.info.emit(f"æ¥æ”¶æ–‡ä»¶å¤±è´¥ï¼š{e}")

                else:
                    # å¯èƒ½æ˜¯ {"ok":true,"msg":"å·²ä¸Šçº¿"} è¿™ç§ç¡®è®¤åŒ…ï¼Œå¿½ç•¥å³å¯
                    pass

        except Exception as e:
            self.emitter.info.emit(f"æ¥æ”¶çº¿ç¨‹å¼‚å¸¸ï¼š{e}")
        finally:
            self.close_chat_connection()

    # ====== ç”¨æˆ·åˆ—è¡¨åˆ·æ–° ======
    def on_users_update(self, users: list):
        # users: [{"username":"alice","online":True,"ip":"x.x.x.x"}, ...]
        # è®©æ‰€æœ‰ç”¨æˆ·éƒ½å­˜åœ¨äº sessionsï¼ˆç¦»çº¿ä¹Ÿä¿ç•™ä¼šè¯ï¼‰
        for u in users:
            name = (u.get("username") or "").strip()
            if not name:
                continue
            self.sessions.setdefault(name, {"messages": [], "unread": 0, "online": False, "ip": ""})
            self.sessions[name]["online"] = bool(u.get("online", False))
            self.sessions[name]["ip"] = u.get("ip", "") or ""

        if self.chatListWidget is None:
            return

        self.chatListWidget.clear()

        peers = [k for k in self.sessions.keys() if k != self.username]
        peers.sort()

        for peer in peers:
            s = self.sessions[peer]
            unread = s["unread"]
            online = "åœ¨çº¿" if s["online"] else "ç¦»çº¿"
            last = ""
            if s["messages"]:
                last_msg = s["messages"][-1]

                # ===== æ–‡ä»¶æ¶ˆæ¯ =====
                if isinstance(last_msg, dict):
                    last = f"[æ–‡ä»¶] {last_msg.get('filename', '')}"
                else:
                    last = last_msg[1]

                if len(last) > 12:
                    last = last[:12] + "..."

            text = f"{peer}  [{online}]"
            if unread > 0:
                text += f"  ({unread})"
            if last:
                text += f"\n{last}"

            # if online == "åœ¨çº¿":
            item = QListWidgetItem(text)
            item.setData(Qt.UserRole, peer)
            self.chatListWidget.addItem(item)

        if self.current_peer is None and self.chatListWidget.count() > 0:
            self.chatListWidget.setCurrentRow(0)
            self.on_peer_clicked(self.chatListWidget.item(0))

        if self.current_peer:
            for i in range(self.chatListWidget.count()):
                it = self.chatListWidget.item(i)
                if it.data(Qt.UserRole) == self.current_peer:
                    self.chatListWidget.setCurrentRow(i)  # é‡æ–°é«˜äº®
                    break

    def on_peer_clicked(self, item: QListWidgetItem):
        peer = item.data(Qt.UserRole)
        self.current_peer = peer

        if peer in self.sessions:
            self.sessions[peer]["unread"] = 0

        self.refresh_message_view(peer)
        # åˆ·æ–°å·¦ä¾§ï¼ˆæ›´æ–°æœªè¯»æ˜¾ç¤ºï¼‰
        self.on_users_update([
            {"username": k, "online": v["online"], "ip": v["ip"]}
            for k, v in self.sessions.items()
        ])

    # ====== æ¶ˆæ¯æ°”æ³¡æ˜¾ç¤º ======
    def refresh_message_view(self, peer: str):
        if self.messageListWidget is None:
            return
        self.messageListWidget.clear()
        msgs = self.sessions.get(peer, {}).get("messages", [])
        for m in msgs:
            if isinstance(m, dict):
                self.add_message_bubble(m)
            else:
                is_me, text = m
                self.add_message_bubble(text, is_me)

    def add_message_bubble(self, msg, is_me=None):
        item = QListWidgetItem()
        item.setFlags(item.flags() ^ Qt.ItemIsSelectable)

        w = QWidget()
        layout = QHBoxLayout(w)
        layout.setContentsMargins(10, 2, 10, 2)

        # ===== æ–‡ä»¶æ¶ˆæ¯ =====
        if isinstance(msg, dict) and msg.get("type") == "file":
            filename = msg["filename"]
            path = msg["path"]
            is_me = msg["is_me"]

            bubble = QLabel(f"ğŸ“„ {filename}\nåŒå‡»æ‰“å¼€")
            bubble.setWordWrap(True)
            bubble.setStyleSheet("""
                QLabel{
                    padding:10px;
                    border-radius:10px;
                    background:#E8F0FE;
                    border:1px solid #C3D3F5;
                }
            """)

            # â­ å…³é”®ï¼šæŠŠæ–‡ä»¶è·¯å¾„ç»‘åœ¨ item ä¸Š
            print("path:", path)
            item.setData(Qt.UserRole, path)


        # ===== æ™®é€šæ–‡æœ¬æ¶ˆæ¯ =====
        else:
            text = msg if isinstance(msg, str) else msg[1]
            bubble = QLabel(text)
            bubble.setWordWrap(True)
            bubble.setMaximumWidth(360)
            bubble.setStyleSheet("""
                QLabel{
                    padding:8px 10px;
                    border-radius:10px;
                    background:#DCF8C6;
                }
            """ if is_me else """
                QLabel{
                    padding:8px 10px;
                    border-radius:10px;
                    background:#FFFFFF;
                    border:1px solid #E5E5E5;
                }
            """)

        if is_me:
            layout.addStretch()
            layout.addWidget(bubble)
        else:
            layout.addWidget(bubble)
            layout.addStretch()

        now = datetime.now()

        # åˆ¤æ–­æ˜¯å¦éœ€è¦æ˜¾ç¤ºæ—¶é—´
        if self.last_msg_time is None or (now - self.last_msg_time).total_seconds() >= 120:
            time_str = now.strftime("%H:%M")
            self.add_time_item(time_str)

        self.last_msg_time = now

        self.messageListWidget.addItem(item)
        self.messageListWidget.setItemWidget(item, w)
        item.setSizeHint(w.sizeHint())
        self.messageListWidget.scrollToBottom()

    # ====== æ”¶åˆ°æ¶ˆæ¯ ======
    def on_message_in(self, _from: str, _to: str, data):
        if isinstance(data, dict) and data.get("type") == "file":
            text = f"[æ–‡ä»¶] {data.get('filename')}"
            msg = {
                "type": "file",
                "is_me": False,
                "filename": data.get('filename'),
                "path": data.get('path'),
            }
        else:
            text = data

        peer = _from if _from != self.username else _to
        self.sessions.setdefault(peer, {"messages": [], "unread": 0, "online": True, "ip": ""})

        is_me = (_from == self.username)
        self.sessions[peer]["messages"].append((is_me, text))

        if self.current_peer == peer:
            if isinstance(data, dict) and data.get("type") == "file":
                self.add_message_bubble(msg, is_me)
            else:
                self.add_message_bubble(text, is_me)
        else:
            self.sessions[peer]["unread"] += 1

        self.on_users_update([
            {"username": k, "online": v["online"], "ip": v["ip"]}
            for k, v in self.sessions.items()
        ])

    # ====== å‘é€æ¶ˆæ¯ ======
    def send_chat_message(self):
        if self.sock is None:
            self.popUp("å‘é€å¤±è´¥", "æœªè¿æ¥æœåŠ¡å™¨", ok=False)
            return
        if self.current_peer is None:
            self.popUp("å‘é€å¤±è´¥", "è¯·å…ˆåœ¨å·¦ä¾§é€‰æ‹©ä¸€ä¸ªç”¨æˆ·/ä¼šè¯", ok=False)
            return
        if self.inputTextEdit is None:
            return

        text = self.inputTextEdit.toPlainText().strip()
        if not text:
            return

        try:
            send_json(self.sock, {
                "action": "msg",
                "from": self.username,
                "to": self.current_peer,
                "text": text,
                "client_ip": self.local_ip
            })

            # æœ¬åœ°å…ˆæ˜¾ç¤ºï¼ˆç«‹å³å‡ºç°ï¼‰
            self.on_message_in(self.username, self.current_peer, text)
            self.inputTextEdit.clear()

        except Exception as e:
            self.popUp("å‘é€å¤±è´¥", str(e), ok=False)

    def send_file(self):
        if not self.sock or not self.current_peer:
            self.popUp("å‘é€å¤±è´¥", "æœªé€‰æ‹©èŠå¤©å¯¹è±¡", ok=False)
            return

        path, _ = QFileDialog.getOpenFileName(
            self.window, "é€‰æ‹©è¦å‘é€çš„æ–‡ä»¶", "", "All Files (*)"
        )
        if not path:
            return

        filename = os.path.basename(path)
        filesize = os.path.getsize(path)

        try:
            # 1ï¸âƒ£ å…ˆå‘ meta
            send_json(self.sock, {
                "action": "file_meta",
                "from": self.username,
                "to": self.current_peer,
                "filename": filename,
                "filesize": filesize
            })

            # 2ï¸âƒ£ å†å‘æ–‡ä»¶å†…å®¹
            with open(path, "rb") as f:
                while True:
                    chunk = f.read(4096)
                    if not chunk:
                        break
                    self.sock.sendall(chunk)

            # æœ¬åœ°æ˜¾ç¤º
            msg = {
                "type": "file",
                "is_me": True,
                "filename": filename,
                "path": path
            }
            self.sessions[self.current_peer]["messages"].append(msg)
            self.add_message_bubble(msg)

        except Exception as e:
            self.popUp("å‘é€å¤±è´¥", str(e), ok=False)

    def open_file(self, item: QListWidgetItem):
        path = item.data(Qt.UserRole)
        if path == None:
            return
        if path and os.path.exists(path):
            QDesktopServices.openUrl(QUrl.fromLocalFile(path))
        else:
            QMessageBox.warning(self.window, "æ— æ³•æ‰“å¼€", "æ–‡ä»¶ä¸å­˜åœ¨æˆ–å·²è¢«åˆ é™¤")

    def add_time_item(self, time_text):
        item = QListWidgetItem()

        label = QLabel(time_text)
        label.setAlignment(Qt.AlignCenter)
        label.setStyleSheet("""
            QLabel {
                color: #888;
                background: transparent;
                padding: 4px;
            }
        """)

        item.setSizeHint(label.sizeHint())
        self.messageListWidget.addItem(item)
        self.messageListWidget.setItemWidget(item, label)


app = QApplication(sys.argv)
w = MainWindow()
w.window.show()
sys.exit(app.exec())
